File: routes/expenses.js
const express = require('express');
const router = express.Router();
const db = require('../database');
const { buildExpensesWhereClause, buildStatsWhereClause, rowToExpense } = require('../helpers/queryBuilders');

router.get('/expenses', async (req, res) => {
  try {
    const limit = req.query.limit != null ? parseInt(req.query.limit, 10) : null;
    const offset = req.query.offset != null ? parseInt(req.query.offset, 10) : 0;

    const parseList = (val) => {
      if (val == null) return undefined;
      if (Array.isArray(val)) return val.filter(Boolean);
      return val.split(',').map(s => s.trim()).filter(Boolean);
    };
    const filters = {
      dateFrom: req.query.dateFrom || undefined,
      dateTo: req.query.dateTo || undefined,
      userId: req.query.userId || undefined,
      categories: parseList(req.query.categories),
      types: parseList(req.query.types),
      labels: parseList(req.query.labels),
      sources: parseList(req.query.sources),
      minAmount: req.query.minAmount,
      maxAmount: req.query.maxAmount,
      search: req.query.search
    };

    const { whereSql, params } = buildExpensesWhereClause(filters);
    const orderBy = 'ORDER BY date DESC';

    if (limit != null && limit >= 0) {
      const countResult = await db.query(
        `SELECT COUNT(*)::int AS total FROM transactions ${whereSql}`,
        params
      );
      const total = countResult.rows[0].total;

      const result = await db.query(
        `SELECT * FROM transactions ${whereSql} ${orderBy} LIMIT $${params.length + 1} OFFSET $${params.length + 2}`,
        [...params, limit, offset]
      );
      const expenses = result.rows.map(rowToExpense);

      return res.json({ expenses, total });
    }

    const result = await db.query(
      `SELECT * FROM transactions ${whereSql} ${orderBy}`,
      params
    );
    const expenses = result.rows.map(rowToExpense);

    res.json(expenses);
  } catch (error) {
    console.error('Error reading expenses:', error);
    res.status(500).json({ error: 'Failed to read expenses' });
  }
});

router.get('/stats', async (req, res) => {
  try {
    const dateFrom = req.query.dateFrom || undefined;
    const dateTo = req.query.dateTo || undefined;
    const userId = req.query.userId || undefined;
    const { whereSql, params } = buildStatsWhereClause(dateFrom, dateTo, userId);

    const baseTable = `(SELECT id, date, type, amount, category, description, user_id FROM transactions ${whereSql}) filtered`;
    const totalResult = await db.query(
      `SELECT
        COALESCE(SUM(CASE WHEN type = 'expense' THEN ABS(amount) ELSE 0 END), 0)::float AS total_expenses,
        COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END), 0)::float AS total_income
      FROM ${baseTable}`,
      params
    );
    const categoryResult = await db.query(
      `SELECT category, COALESCE(SUM(ABS(amount)), 0)::float AS total
       FROM ${baseTable} WHERE type = 'expense' GROUP BY category`,
      params
    );
    const incomeCategoryResult = await db.query(
      `SELECT category, COALESCE(SUM(amount), 0)::float AS total
       FROM ${baseTable} WHERE type = 'income' GROUP BY category`,
      params
    );
    const monthlyResult = await db.query(
      `SELECT
        to_char(date, 'YYYY-MM') AS iso_month,
        to_char(date, 'Mon YYYY') AS display_month,
        date_trunc('month', date) AS month_start,
        COALESCE(SUM(CASE WHEN type = 'expense' THEN ABS(amount) ELSE 0 END), 0)::float AS expenses,
        COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END), 0)::float AS income
       FROM ${baseTable}
       GROUP BY to_char(date, 'YYYY-MM'), to_char(date, 'Mon YYYY'), date_trunc('month', date)
       ORDER BY month_start`,
      params
    );
    const monthlyCategoryResult = await db.query(
      `SELECT
        to_char(date, 'Mon YYYY') AS display_month,
        date_trunc('month', date) AS month_start,
        category,
        COALESCE(SUM(ABS(amount)), 0)::float AS total
       FROM ${baseTable} WHERE type = 'expense'
       GROUP BY to_char(date, 'Mon YYYY'), date_trunc('month', date), category
       ORDER BY month_start`,
      params
    );
    const topExpensesResult = await db.query(
      `SELECT id, date, description, category, amount, type FROM ${baseTable}
       WHERE type = 'expense' ORDER BY ABS(amount) DESC LIMIT 10`,
      params
    );
    const topIncomeResult = await db.query(
      `SELECT id, date, description, category, amount, type FROM ${baseTable}
       WHERE type = 'income' ORDER BY amount DESC LIMIT 10`,
      params
    );

    const totalExpenses = parseFloat(totalResult.rows[0].total_expenses) || 0;
    const totalIncome = parseFloat(totalResult.rows[0].total_income) || 0;
    const categoryBreakdown = {};
    categoryResult.rows.forEach(row => {
      const cat = row.category || 'Uncategorized';
      categoryBreakdown[cat] = parseFloat(row.total) || 0;
    });
    const incomeCategoryBreakdown = {};
    incomeCategoryResult.rows.forEach(row => {
      const cat = row.category || 'Uncategorized';
      incomeCategoryBreakdown[cat] = parseFloat(row.total) || 0;
    });
    const monthlyData = monthlyResult.rows.map(row => ({
      month: row.display_month,
      expenses: parseFloat(row.expenses) || 0,
      income: parseFloat(row.income) || 0
    }));
    const monthOrder = monthlyResult.rows.map(r => r.display_month);
    const monthlyCategoryMap = new Map();
    monthlyCategoryResult.rows.forEach(row => {
      const key = row.display_month;
      if (!monthlyCategoryMap.has(key)) {
        monthlyCategoryMap.set(key, { month: key });
      }
      monthlyCategoryMap.get(key)[row.category || 'Uncategorized'] = parseFloat(row.total) || 0;
    });
    const monthlyCategoryData = monthOrder.map(m => monthlyCategoryMap.get(m) || { month: m });

    const topExpenses = topExpensesResult.rows.map(row => ({
      id: row.id,
      date: row.date,
      description: row.description,
      category: row.category,
      amount: parseFloat(row.amount),
      type: 'expense',
      user: row.user_id || ''
    }));
    const topIncome = topIncomeResult.rows.map(row => ({
      id: row.id,
      date: row.date,
      description: row.description,
      category: row.category,
      amount: parseFloat(row.amount),
      type: 'income',
      user: row.user_id || ''
    }));

    res.json({
      totalExpenses,
      totalIncome,
      netAmount: totalIncome - totalExpenses,
      categoryBreakdown,
      incomeCategoryBreakdown,
      monthlyData,
      monthlyCategoryData,
      topExpenses,
      topIncome
    });
  } catch (error) {
    console.error('Error reading stats:', error);
    res.status(500).json({ error: 'Failed to read stats' });
  }
});

router.patch('/expenses/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const expenseData = req.body;

    const fields = [];
    const values = [];
    let paramIndex = 1;

    if (expenseData.date) {
      fields.push(`date = $${paramIndex++}`);
      values.push(expenseData.date);
    }
    if (expenseData.description) {
      fields.push(`description = $${paramIndex++}`);
      values.push(expenseData.description);
    }
    if (expenseData.hasOwnProperty('category')) {
      fields.push(`category = $${paramIndex++}`);
      values.push(expenseData.category || null);
    }
    if (expenseData.amount) {
      fields.push(`amount = $${paramIndex++}`);
      values.push(expenseData.amount);
    }
    if (expenseData.type) {
      fields.push(`type = $${paramIndex++}`);
      values.push(expenseData.type);
    }
    if (expenseData.user) {
      fields.push(`user_id = $${paramIndex++}`);
      values.push(expenseData.user);
    }
    if (expenseData.labels) {
      fields.push(`labels = $${paramIndex++}`);
      values.push(JSON.stringify(expenseData.labels));
    }
    if (expenseData.excludedFromCalculations !== undefined) {
      fields.push(`excluded_from_calculations = $${paramIndex++}`);
      values.push(expenseData.excludedFromCalculations);
    }

    if (fields.length === 0) {
      return res.status(400).json({ error: 'No fields to update' });
    }

    fields.push('updated_at = CURRENT_TIMESTAMP');
    const setClause = fields.join(', ');
    values.push(id);

    const query = `UPDATE transactions SET ${setClause} WHERE id = $${paramIndex}`;
    const result = await db.query(query, values);

    if (result.rowCount === 0) {
      return res.status(404).json({ error: 'Transaction not found' });
    }

    const updatedResult = await db.query('SELECT * FROM transactions WHERE id = $1', [id]);
    const updatedExpense = rowToExpense(updatedResult.rows[0]);

    res.json(updatedExpense);
  } catch (error) {
    console.error('Error updating expense:', error);
    res.status(500).json({ error: 'Failed to update expense' });
  }
});

router.post('/expenses', async (req, res) => {
  try {
    const { expenses, metadata } = req.body;

    const client = await db.beginTransaction();
    try {
      await client.query('DELETE FROM transactions');

      for (const expense of expenses) {
        await client.query(
          `INSERT INTO transactions (
            id, date, description, category, amount, type, user_id,
            labels, metadata, transfer_info, excluded_from_calculations
          ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
          ON CONFLICT (id) DO UPDATE SET
            date = EXCLUDED.date,
            description = EXCLUDED.description,
            category = EXCLUDED.category,
            amount = EXCLUDED.amount,
            type = EXCLUDED.type,
            user_id = EXCLUDED.user_id,
            labels = EXCLUDED.labels,
            metadata = EXCLUDED.metadata,
            transfer_info = EXCLUDED.transfer_info,
            excluded_from_calculations = EXCLUDED.excluded_from_calculations,
            updated_at = CURRENT_TIMESTAMP`,
          [
            expense.id,
            expense.date,
            expense.description,
            expense.category || 'Uncategorized',
            expense.amount,
            expense.type,
            expense.user,
            JSON.stringify(expense.labels || []),
            JSON.stringify(expense.metadata || {}),
            expense.transferInfo ? JSON.stringify(expense.transferInfo) : null,
            expense.excludedFromCalculations || false
          ]
        );
      }

      if (metadata) {
        await client.query(
          'INSERT INTO metadata (key, value) VALUES ($1, $2) ON CONFLICT (key) DO UPDATE SET value = EXCLUDED.value',
          ['storage_metadata', JSON.stringify(metadata)]
        );
      }

      await db.commitTransaction(client);

      res.json({ success: true, count: expenses.length });
    } catch (error) {
      await db.rollbackTransaction(client);
      throw error;
    }
  } catch (error) {
    console.error('Error saving expenses:', error);
    res.status(500).json({ error: 'Failed to save expenses' });
  }
});

module.exports = router;

File: backend/app/routes/expenses.py
from datetime import date as date_type
from decimal import Decimal

from fastapi import APIRouter, BackgroundTasks, Body, Depends, Query
from fastapi.responses import JSONResponse
from sqlalchemy import case, delete, func, select, text
from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.database import get_db
from app.demo.limits import assert_demo_not_mass_delete, assert_demo_replace_count
from app.dependencies.auth import get_current_household_id
from app.models.metadata import Metadata
from app.models.transaction import Transaction
from app.schemas.transaction import (
    ExpenseBulkSaveRequest,
    TransactionOut,
    TransactionUpdate,
)
from app.utils.query_builder import build_expenses_filter, build_stats_filter
from app.utils.subscription_signature import normalize_signature
from app.utils.subscription_utils import reconcile_signature_bg, run_detection_bg

router = APIRouter(prefix="/api", tags=["expenses"])


def _parse_list(value: str | None) -> list[str] | None:
    """Parse comma-separated query param into list."""
    if not value:
        return None
    return [s.strip() for s in value.split(",") if s.strip()]


@router.get("/expenses")
async def get_expenses(
    limit: int | None = Query(None),
    offset: int = Query(0),
    dateFrom: str | None = Query(None),
    dateTo: str | None = Query(None),
    userId: str | None = Query(None),
    categories: str | None = Query(None),
    types: str | None = Query(None),
    labels: str | None = Query(None),
    sources: str | None = Query(None),
    minAmount: str | None = Query(None),
    maxAmount: str | None = Query(None),
    search: str | None = Query(None),
    household_id: str = Depends(get_current_household_id),
    db: AsyncSession = Depends(get_db),
):
    filter_params = {
        "dateFrom": dateFrom,
        "dateTo": dateTo,
        "userId": userId,
        "categories": _parse_list(categories),
        "types": _parse_list(types),
        "labels": _parse_list(labels),
        "sources": _parse_list(sources),
        "minAmount": minAmount,
        "maxAmount": maxAmount,
        "search": search,
    }
    filters = [Transaction.household_id == household_id] + build_expenses_filter(
        filter_params
    )
    base_stmt = select(Transaction).where(*filters)
    order = base_stmt.order_by(Transaction.date.desc())

    if limit is not None and limit >= 0:
        count_stmt = select(func.count()).select_from(base_stmt.subquery())
        total = (await db.execute(count_stmt)).scalar()

        result = await db.execute(order.limit(limit).offset(offset))
        expenses = [
            TransactionOut.from_orm_model(t).model_dump()
            for t in result.scalars().all()
        ]
        return {"expenses": expenses, "total": total}

    result = await db.execute(order)
    return [
        TransactionOut.from_orm_model(t).model_dump() for t in result.scalars().all()
    ]


@router.get("/stats")
async def get_stats(
    dateFrom: str | None = Query(None),
    dateTo: str | None = Query(None),
    userId: str | None = Query(None),
    household_id: str = Depends(get_current_household_id),
    db: AsyncSession = Depends(get_db),
):
    filters = [Transaction.household_id == household_id] + build_stats_filter(
        dateFrom, dateTo, userId
    )
    base = (
        select(
            Transaction.id,
            Transaction.date,
            Transaction.type,
            Transaction.amount,
            Transaction.category,
            Transaction.description,
            Transaction.created_by_user_id,
            Transaction.household_id,
        )
        .where(*filters)
        .subquery()
    )

    # Totals
    total_row = (
        await db.execute(
            select(
                func.coalesce(
                    func.sum(
                        case(
                            (base.c.type == "expense", func.abs(base.c.amount)),
                            else_=0,
                        )
                    ),
                    0,
                ).label("total_expenses"),
                func.coalesce(
                    func.sum(case((base.c.type == "income", base.c.amount), else_=0)),
                    0,
                ).label("total_income"),
            ).select_from(base)
        )
    ).one()
    total_expenses = float(total_row.total_expenses)
    total_income = float(total_row.total_income)

    # Category breakdown (expenses)
    cat_rows = (
        await db.execute(
            select(
                base.c.category,
                func.coalesce(func.sum(func.abs(base.c.amount)), 0).label("total"),
            )
            .select_from(base)
            .where(base.c.type == "expense")
            .group_by(base.c.category)
        )
    ).all()
    category_breakdown = {
        (row.category or "Uncategorized"): float(row.total) for row in cat_rows
    }

    # Income category breakdown
    income_cat_rows = (
        await db.execute(
            select(
                base.c.category,
                func.coalesce(func.sum(base.c.amount), 0).label("total"),
            )
            .select_from(base)
            .where(base.c.type == "income")
            .group_by(base.c.category)
        )
    ).all()
    income_category_breakdown = {
        (row.category or "Uncategorized"): float(row.total) for row in income_cat_rows
    }

    # Monthly data — use GROUP BY position to avoid duplicate bind params
    monthly_rows = (
        await db.execute(
            select(
                func.to_char(base.c.date, "YYYY-MM").label("iso_month"),
                func.to_char(base.c.date, "Mon YYYY").label("display_month"),
                func.date_trunc("month", base.c.date).label("month_start"),
                func.coalesce(
                    func.sum(
                        case(
                            (base.c.type == "expense", func.abs(base.c.amount)),
                            else_=0,
                        )
                    ),
                    0,
                ).label("expenses"),
                func.coalesce(
                    func.sum(case((base.c.type == "income", base.c.amount), else_=0)),
                    0,
                ).label("income"),
            )
            .select_from(base)
            .group_by(text("1, 2, 3"))
            .order_by(text("month_start"))
        )
    ).all()
    monthly_data = [
        {
            "month": row.display_month,
            "expenses": float(row.expenses),
            "income": float(row.income),
        }
        for row in monthly_rows
    ]
    month_order = [row.display_month for row in monthly_rows]

    # Monthly category data
    monthly_cat_rows = (
        await db.execute(
            select(
                func.to_char(base.c.date, "Mon YYYY").label("display_month"),
                func.date_trunc("month", base.c.date).label("month_start"),
                base.c.category,
                func.coalesce(func.sum(func.abs(base.c.amount)), 0).label("total"),
            )
            .select_from(base)
            .where(base.c.type == "expense")
            .group_by(text("1, 2, 3"))
            .order_by(text("month_start"))
        )
    ).all()
    monthly_cat_map: dict[str, dict] = {}
    for row in monthly_cat_rows:
        key = row.display_month
        if key not in monthly_cat_map:
            monthly_cat_map[key] = {"month": key}
        monthly_cat_map[key][row.category or "Uncategorized"] = float(row.total)
    monthly_category_data = [monthly_cat_map.get(m, {"month": m}) for m in month_order]

    # Top 10 expenses
    top_exp_rows = (
        await db.execute(
            select(
                base.c.id,
                base.c.date,
                base.c.description,
                base.c.category,
                base.c.amount,
                base.c.created_by_user_id,
            )
            .select_from(base)
            .where(base.c.type == "expense")
            .order_by(func.abs(base.c.amount).desc())
            .limit(10)
        )
    ).all()
    top_expenses = [
        {
            "id": r.id,
            "date": r.date.isoformat() if hasattr(r.date, "isoformat") else r.date,
            "description": r.description,
            "category": r.category,
            "amount": float(r.amount),
            "type": "expense",
            "user": r.created_by_user_id or "",
        }
        for r in top_exp_rows
    ]

    # Top 10 income
    top_inc_rows = (
        await db.execute(
            select(
                base.c.id,
                base.c.date,
                base.c.description,
                base.c.category,
                base.c.amount,
                base.c.created_by_user_id,
            )
            .select_from(base)
            .where(base.c.type == "income")
            .order_by(base.c.amount.desc())
            .limit(10)
        )
    ).all()
    top_income = [
        {
            "id": r.id,
            "date": r.date.isoformat() if hasattr(r.date, "isoformat") else r.date,
            "description": r.description,
            "category": r.category,
            "amount": float(r.amount),
            "type": "income",
            "user": r.created_by_user_id or "",
        }
        for r in top_inc_rows
    ]

    return {
        "totalExpenses": total_expenses,
        "totalIncome": total_income,
        "netAmount": total_income - total_expenses,
        "categoryBreakdown": category_breakdown,
        "incomeCategoryBreakdown": income_category_breakdown,
        "monthlyData": monthly_data,
        "monthlyCategoryData": monthly_category_data,
        "topExpenses": top_expenses,
        "topIncome": top_income,
    }


@router.patch("/expenses/{expense_id}")
async def update_expense(
    expense_id: str,
    bg: BackgroundTasks,
    body: TransactionUpdate = Body(...),
    household_id: str = Depends(get_current_household_id),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Transaction).where(
            Transaction.id == expense_id, Transaction.household_id == household_id
        )
    )
    txn = result.scalar_one_or_none()
    if not txn:
        return JSONResponse(status_code=404, content={"error": "Transaction not found"})

    old_description = txn.description

    updated = False
    if body.date is not None:
        txn.date = body.date
        updated = True
    if body.description is not None:
        txn.description = body.description
        updated = True
    if body.category is not None:
        txn.category = body.category
        updated = True
    if body.amount is not None:
        txn.amount = Decimal(str(body.amount))
        updated = True
    if body.type is not None:
        txn.type = body.type
        updated = True
    if body.user is not None:
        txn.created_by_user_id = body.user
        updated = True
    if body.labels is not None:
        txn.labels = body.labels
        updated = True
    if body.excludedFromCalculations is not None:
        txn.excluded_from_calculations = body.excludedFromCalculations
        updated = True
    if body.transferInfo is not None:
        txn.transfer_info = body.transferInfo
        updated = True

    if not updated:
        return JSONResponse(status_code=400, content={"error": "No fields to update"})

    await db.commit()
    await db.refresh(txn)

    old_sig = normalize_signature(old_description)
    new_sig = normalize_signature(txn.description)
    bg.add_task(reconcile_signature_bg, household_id, new_sig)
    if old_sig and old_sig != new_sig:
        bg.add_task(reconcile_signature_bg, household_id, old_sig)

    return TransactionOut.from_orm_model(txn).model_dump()


@router.post("/expenses")
async def bulk_save_expenses(
    body: ExpenseBulkSaveRequest,
    bg: BackgroundTasks,
    household_id: str = Depends(get_current_household_id),
    db: AsyncSession = Depends(get_db),
):
    assert_demo_replace_count(
        len(body.expenses),
        cap=settings.demo_max_transactions,
        entity="transactions",
    )
    await assert_demo_not_mass_delete(db, Transaction, household_id, len(body.expenses))

    # Delete existing transactions for this household only.
    await db.execute(
        delete(Transaction).where(Transaction.household_id == household_id)
    )

    # Insert new transactions, attached to the requested household.
    for exp in body.expenses:
        txn = Transaction(
            id=exp.id,
            date=date_type.fromisoformat(str(exp.date)[:10]),
            description=exp.description,
            category=exp.category or "Uncategorized",
            amount=exp.amount,
            type=exp.type,
            household_id=household_id,
            created_by_user_id=exp.user,
            labels=exp.labels or [],
            metadata_=exp.metadata or {},
            transfer_info=exp.transferInfo,
            excluded_from_calculations=exp.excludedFromCalculations or False,
        )
        db.add(txn)

    # Store metadata if provided
    if body.metadata:
        stmt = (
            insert(Metadata)
            .values(key="storage_metadata", value=body.metadata)
            .on_conflict_do_update(
                index_elements=[Metadata.key],
                set_={"value": body.metadata},
            )
        )
        await db.execute(stmt)

    await db.commit()
    bg.add_task(run_detection_bg, household_id)

    return {"success": True, "count": len(body.expenses)}
